crates/adapters/src/controller/stats.rs

```diff
@@ -1310,6 +1410,232 @@ impl TokenList {
+/// has been processed by the pipeline.
+#[repr(transparent)]
+pub(crate) struct WatermarkTracker(Mutex<WatermarkTrackerInner>);
+
+impl WatermarkTracker {
+    fn new() -> Self {
+        Self(Mutex::new(WatermarkTrackerInner::new()))
+    }
+
+    fn append(&self, watermarks: Vec<Watermark>, global_offset: u64) {
+        self.0.lock().unwrap().append(watermarks, global_offset);
+    }
+
+    /// Called when `total_processed_records` changes. Updates processed_at timestamps.
+    fn update_processed(&self, processed_records: u64, ts: DateTime<Utc>) {
+        self.0
+            .lock()
+            .unwrap()
+            .update_processed(processed_records, ts);
+    }
+
+    /// Called when `total_completed_records` changes. Updates histograms, `completed_watermark`,
+    /// and removes completed watermarks from the list.
+    fn update_completed(
+        &self,
+        completed_records: u64,
+        ts: DateTime<Utc>,
+        metrics: &InputEndpointMetrics,
+    ) {
+        self.0
+            .lock()
+            .unwrap()
+            .update_completed(completed_records, ts, metrics);
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn debug(&self) -> String {
+        format!("{:?}", self.0.lock().unwrap())
+    }
+
+    #[allow(dead_code)]
+    pub fn completed_watermark(&self) -> Option<CompletedWatermark> {
+        self.0.lock().unwrap().completed_watermark.clone()
+    }
+}
+
+/// Bound the number of tracked watermarks to avoid unbounded memory growth.
+const MAX_TRACKED_WATERMARKS: usize = 10_000;
+
+#[derive(Debug)]
+struct WatermarkTrackerInner {
+    /// The last `total_completed_records` value reported by the controller.
+    total_completed_records: u64,
+
+    /// The last `total_processed_records` value reported by the controller.
+    total_processed_records: u64,
+
+    /// Not-yet-completed watermarks, in the order they were added.
+    watermark_list: VecDeque<WatermarkListEntry>,
+
+    /// The latest completed watermark with non-empty metadata, if any.
+    completed_watermark: Option<CompletedWatermark>,
+}
+
+impl WatermarkTrackerInner {
+    fn new() -> Self {
+        Self {
+            total_completed_records: 0,
+            total_processed_records: 0,
+            watermark_list: VecDeque::new(),
+            completed_watermark: None,
+        }
+    }
+
+    fn append(&mut self, watermarks: Vec<Watermark>, global_offset: u64) {
+        for watermark in watermarks {
+            self.watermark_list.push_back(WatermarkListEntry {
+                watermark,
+                global_offset,
+                processed_at: None,
+            });
+        }
+
+        // Truncate newly added watermarks to keep the queue bounded. This gives older watermarks
+        // a chance to complete; otherwise we risk always evicting all watermarks before the complete
+        // and never reporting any completed watermarks.
+        self.watermark_list.truncate(MAX_TRACKED_WATERMARKS);
```

Comment:
you could just not add them, that would be cheaper

Reply:
fair